src/exodus/sigmawallet.cpp

```diff
@@ -0,0 +1,597 @@
+#include "exodus.h"
+
+#include "../main.h"
+#include "../init.h"
+
+#include "../sigma/openssl_context.h"
+#include "../crypto/hmac_sha256.h"
+#include "../crypto/hmac_sha512.h"
+
+#include <boost/optional.hpp>
+
+namespace exodus
+{
+
+MintPoolEntry::MintPoolEntry()
+{
+}
+
+MintPoolEntry::MintPoolEntry(SigmaPublicKey const &key, CKeyID const &seedId)
+    : key(key), seedId(seedId)
+{
+}
+
+bool MintPoolEntry::operator==(MintPoolEntry const &another) const
+{
+    return key == another.key && seedId == another.seedId;
+}
+
+bool MintPoolEntry::operator!=(MintPoolEntry const &another) const
+{
+    return !(*this == another);
+}
+
+SigmaWallet::SigmaWallet(const std::string& walletFile) : walletFile(walletFile)
+{
+    // Don't try to do anything else if the wallet is locked.
+    if (pwalletMain->IsLocked()) {
+        return;
+    }
+
+    // Use MasterKeyId from HDChain as index for mintpool
+    auto masterId = pwalletMain->GetHDChain().masterKeyID;
+    LogPrintf("masterId: %d\n", masterId.GetHex());
+
+    if (!SetupWallet(masterId)) {
+
+        LogPrintf("%s: failed to save deterministic seed for hashseed %s\n", __func__, masterId.GetHex());
+        throw std::runtime_error("fail to setup wallet");
+    }
+}
+
+bool SigmaWallet::SetupWallet(const uint160& masterId)
+{
+    CWalletDB walletdb(walletFile);
+
+    if (pwalletMain->IsLocked()) {
+        return false;
+    }
+
+    if (masterId.IsNull()) {
+        return error("%s: failed to set master seed.", __func__);
+    }
+
+    this->masterId = masterId;
+
+    // Load mint pool from DB
+    LoadMintPool();
+
+    // Clean up any mint entry that isn't corresponded to current masterId
+    CleanUp();
+
+    // Refill mint pool
+    GenerateMintPool();
+
+    return true;
+}
+
+// Generator
+uint32_t SigmaWallet::CreateNextSeed(CKeyID &seedId, uint512& seed)
+{
+    LOCK(pwalletMain->cs_wallet);
+    seedId = pwalletMain->GenerateNewKey(BIP44_EXODUS_MINT_INDEX).GetID();
+    return GenerateSeed(seedId, seed);
+}
+
+uint32_t SigmaWallet::GenerateSeed(CKeyID const &seedId, uint512& seed)
+{
+    CKey key;
+    if (!pwalletMain->CCryptoKeyStore::GetKey(seedId, key)) {
+        throw std::runtime_error(
+            "Unable to retrieve generated key for mint seed. Is the wallet locked?");
+    }
+
+    // HMAC-SHA512(SHA256(count), key)
+    std::array<unsigned char, CSHA256::OUTPUT_SIZE> countHash;
+
+    std::vector<unsigned char> result;
+    result.resize(CSHA512::OUTPUT_SIZE);
+
+    auto count = std::to_string(GetSeedIndex(seedId));
+    CSHA256().
+        Write(
+            reinterpret_cast<const unsigned char*>(count.data()),
+            count.size()
+        ).
+        Finalize(countHash.data());
+
+    CHMAC_SHA512(countHash.data(), sizeof(countHash)).
+        Write(key.begin(), key.size()).
+        Finalize(result.data());
+
+    seed = uint512(result);
+
+    return GetSeedIndex(seedId);
+}
+
+bool GetChildFromPath(string const &path, std::pair<uint32_t, bool> &child)
```

Comment:
IMO we should use regular expression instead. The benefits are:

- Easy to read what format we are dealing with.
- Less error prone.